### PR TITLE
Refactor lifecycle tests.

### PR DESCRIPTION
**Pull request by pgavlin**
Move these tests to a new package, `lifecycletest`, that also exposes
APIs that allow consumers to implement their own lifecycle tests. This
is intended to ease the burden of testing plugin implementations and to
set the stage for cleaning up the lifecycle tests themselves.

This involves two changes to the public API, only one of which is
strictly necessary:

- The `host` field of `UpdateOptions` is now exported
- The `Journal` type has been moved from test-only code to the package
  proper

The former change is necessary, as it is the mechanism by which package
consumers may inject their own plugin loaders. I was reluctant to expose
this field originally because I wanted to ensure that the behavior of
packages that embed Pulumi is consistent with that of the Pulumi CLI
with respect to plugin loading. I now believe that the risk of consumers
changing this behavior outside of test scenarios is low enough that we
can expose this field. This may also be useful for future scenarios,
e.g. statically linking providers and Pulumi programs.

The latter change is not necessary, but fleshes out the engine package
into a more complete toolkit. Downstream consumers may use the Journal
type to conveniently implement snapshotting.

### CODE REVIEW

File: pkg/engine/journal.go
Comment (lblackstone) [inline pkg/engine/journal.go:178]: Is this intentionally a no-op?

File: pkg/engine/lifeycletest/test_plan.go
Comment (lblackstone) [inline pkg/engine/lifeycletest/test_plan.go:69]: Is this channel closed elsewhere?

File: pkg/engine/lifeycletest/test_plan.go
Comment (lblackstone) [inline pkg/engine/lifeycletest/test_plan.go:289]:
```suggestion
				// Should see only refresh-sames.
```

**Review by lblackstone — APPROVED** [review]

File: pkg/engine/journal.go
Comment (pgavlin) [reply pkg/engine/journal.go:178]: Yes.

Comment (pgavlin) [reply pkg/engine/journal.go:178]: (I think)

File: pkg/engine/lifeycletest/test_plan.go
Comment (pgavlin) [reply pkg/engine/lifeycletest/test_plan.go:69]: I believe so. I'll double-check after the fact, as I need to land this PR to unblock some refactoring

**Review by EvanBoyle — COMMENTED** [review]
👍

**Pull request merged**
